rewrite-core/src/main/java/org/openrewrite/PathUtils.java

```diff
@@ -44,17 +49,172 @@ public static boolean equalIgnoringSeparators(String a , String b) {
+                return false;
+            }
+            if (pattIdxEnd == (pattTokens.length - 1)
+                    && (isFileSeparator(pattern.charAt(pattern.length() - 1)) ^ isFileSeparator(path.charAt(path.length() - 1)))) {
+                return false;
+            }
+            pattIdxEnd--;
+            pathIdxEnd--;
+        }
+        if (pathIdxStart > pathIdxEnd) {
+            // Path exhausted
+            for (int i = pattIdxStart; i < pattIdxEnd; i++) {
+                if (!pattTokens[i].equals("**")) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        while (pattIdxStart != pattIdxEnd && pathIdxStart <= pathIdxEnd) {
+            int patIdxTmp = -1;
+            for (int i = pattIdxStart + 1; i <= pattIdxEnd; i++) {
+                if (pattTokens[i].equals("**")) {
+                    patIdxTmp = i;
+                    break;
+                }
+            }
+            if (patIdxTmp == pattIdxStart + 1) {
+                // **/** situation, so skip one
+                pattIdxStart++;
+                continue;
+            }
+            // Find the pattern between pattIdxStart & pattIdxTmp in path between
+            // pathIdxStart & pathIdxEnd
+            int patLength = (patIdxTmp - pattIdxStart - 1);
+            int strLength = (pathIdxEnd - pathIdxStart + 1);
+            int foundIdx = -1;
+            strLoop:
+            for (int i = 0; i <= strLength - patLength; i++) {
+                for (int j = 0; j < patLength; j++) {
+                    if (!StringUtils.matchesGlob(pathTokens[pathIdxStart + i + j], pattTokens[pattIdxStart + j + 1])) {
+                        continue strLoop;
+                    }
+                }
+                foundIdx = pathIdxStart + i;
+                break;
+            }
+
+            if (foundIdx == -1) {
+                return false;
+            }
+            pattIdxStart = patIdxTmp;
+            pathIdxStart = foundIdx + patLength;
+        }
+
+        for (int i = pattIdxStart; i <= pattIdxEnd; i++) {
+            if (!pattTokens[i].equals("**")) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static String[] tokenize(String path) {
+        List<String> tokens = new ArrayList<>();
+        int pathIdxStart = 0;
+        int pathIdxTmp = 0;
+        int pathIdxEnd = path.length() - 1;
+        while (pathIdxTmp <= pathIdxEnd) {
+            if (isFileSeparator(path.charAt(pathIdxTmp))) {
+                tokens.add(path.substring(pathIdxStart, pathIdxTmp));
+                pathIdxStart = pathIdxTmp + 1;
+            }
+            pathIdxTmp++;
+        }
+        if (pathIdxStart == 0) {
+            tokens.add(path);
+        } else if (pathIdxStart != pathIdxTmp) {
+            tokens.add(path.substring(pathIdxStart, pathIdxTmp));
+        }
+        return tokens.toArray(new String[0]);
```

Comment:
You could also use JDK's `StringTokenizer`. Something like:

```java
        List<String> tokens = new ArrayList<>();
        StringTokenizer tokenizer = new StringTokenizer(path, "/\\");
        while (tokenizer.hasMoreTokens()) {
            tokens.add(tokenizer.nextToken());
        }
        return tokens.toArray(new String[0]);
```

Reply:
So I've seen and used this before as well. The one thing to note is that the current docs state that it's only being kept for backward compatibility and it's use is discourage. It is instead recommend using `String#split(String)` (which accepts a regex) or `Pattern.compile(String).split(String)`.

The current implementation was based upon those statement as well as trying to avoid regex which usually carries a higher performance profile.

To be on the safe side, I'll go ahead and setup a JMH test case to measure the performance of each of the variants, then we can go with the most performant of them. Sound good to you?

Reply:
Sounds good. I wasn't even aware that it had been quasi-deprecated 🙂

Reply:
Generated using Lenovo P53s laptop `Windows 10, i7-8665U, 32 GB RAM, Eclipse Temurin JDK 17.0.5+8`

Input paths are 27 tokens long each like so:
```
private static final String UNIX_PATH = "a/b/c/d/e/f/g/h/i/j/k/l/m/n/o/p/q/r/s/t/u/v/w/x/y/z/hello.txt";
private static final String WINDOWS_PATH = UNIX_PATH.replace("/", "\\");
```

```
baseline = currrent charAt implementation
stringTokenizer = new StringTokenizer(path, "/\\\\")
stringSplit = path.split("/\\\\")
patternSplit = PATTERN.split(path) // PATTERN = Pattern.compile("/\\\\"); precompiled outside of benchmark scope
```

Results:
```
Benchmark                                                                            Mode      Cnt       Score   Error  Units
PathTokenizationBenchmark.baseline_UNIX                                            sample  3078277       1.785 ± 0.035  us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.00                        sample                0.800          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.50                        sample                1.800          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.90                        sample                2.000          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.95                        sample                2.100          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.99                        sample                3.100          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.999                       sample               32.288          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p0.9999                      sample              273.937          us/op
PathTokenizationBenchmark.baseline_UNIX:baseline_UNIX·p1.00                        sample            15695.872          us/op
PathTokenizationBenchmark.baseline_WINDOWS                                         sample  2931319       1.846 ± 0.027  us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.00                  sample                0.900          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.50                  sample                1.800          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.90                  sample                2.100          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.95                  sample                2.100          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.99                  sample                3.100          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.999                 sample               37.440          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p0.9999                sample              285.628          us/op
PathTokenizationBenchmark.baseline_WINDOWS:baseline_WINDOWS·p1.00                  sample             6234.112          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX                                     sample  1995709       3.291 ± 0.633  us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.00          sample                0.600          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.50          sample                1.400          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.90          sample                1.500          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.95          sample                1.500          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.99          sample                1.700          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.999         sample              111.616          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p0.9999        sample             2496.221          us/op
PathTokenizationBenchmark.stringTokenizer_UNIX:stringTokenizer_UNIX·p1.00          sample           156237.824          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS                                  sample  2299815       1.236 ± 0.030  us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.00    sample                0.600          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.50    sample                1.100          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.90    sample                1.500          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.95    sample                1.500          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.99    sample                1.600          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.999   sample               25.106          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p0.9999  sample              227.091          us/op
PathTokenizationBenchmark.stringTokenizer_WINDOWS:stringTokenizer_WINDOWS·p1.00    sample             9322.496          us/op
PathTokenizationBenchmark.stringSplit_UNIX                                         sample  2069065       1.361 ± 0.029  us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.00                  sample                0.700          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.50                  sample                1.300          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.90                  sample                1.600          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.95                  sample                1.700          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.99                  sample                1.800          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.999                 sample               31.872          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p0.9999                sample              212.760          us/op
PathTokenizationBenchmark.stringSplit_UNIX:stringSplit_UNIX·p1.00                  sample             6823.936          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS                                      sample  2082495       0.741 ± 0.024  us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.00            sample                0.300          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.50            sample                0.500          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.90            sample                0.900          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.95            sample                0.900          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.99            sample                1.100          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.999           sample               17.792          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p0.9999          sample              142.016          us/op
PathTokenizationBenchmark.stringSplit_WINDOWS:stringSplit_WINDOWS·p1.00            sample             4276.224          us/op
PathTokenizationBenchmark.patternSplit_UNIX                                        sample  2545698       1.158 ± 0.041  us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.00                sample                0.500          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.50                sample                1.100          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.90                sample                1.200          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.95                sample                1.300          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.99                sample                1.500          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.999               sample               30.784          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p0.9999              sample              239.580          us/op
PathTokenizationBenchmark.patternSplit_UNIX:patternSplit_UNIX·p1.00                sample            11534.336          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS                                     sample  1883742       0.420 ± 0.024  us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.00          sample                0.200          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.50          sample                0.300          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.90          sample                0.500          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.95          sample                0.500          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.99          sample                0.500          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.999         sample                9.888          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p0.9999        sample              101.968          us/op
PathTokenizationBenchmark.patternSplit_WINDOWS:patternSplit_WINDOWS·p1.00          sample             7340.032          us/op



Benchmark                                           Mode  Cnt   Score   Error   Units
PathTokenizationBenchmark.baseline_UNIX            thrpt    2   1.653          ops/us
PathTokenizationBenchmark.baseline_WINDOWS         thrpt    2   2.477          ops/us
PathTokenizationBenchmark.stringTokenizer_UNIX     thrpt    2   3.624          ops/us
PathTokenizationBenchmark.stringTokenizer_WINDOWS  thrpt    2   3.095          ops/us
PathTokenizationBenchmark.stringSplit_UNIX         thrpt    2   3.280          ops/us
PathTokenizationBenchmark.stringSplit_WINDOWS      thrpt    2   5.345          ops/us
PathTokenizationBenchmark.patternSplit_UNIX        thrpt    2   3.234          ops/us
PathTokenizationBenchmark.patternSplit_WINDOWS     thrpt    2  11.293          ops/us
```

It seems like `String#split(String)` or `Pattern.split(String)` are going to be the most performant ways to tokenize the output. Coincidentally enough their implementations are also the simplest.

Reply:
Now, sadly it doesn't seem to keep up when it's actually put into the glob matching code itself.
```
BEFORE: (charAt)
Benchmark                                          Mode  Cnt  Score   Error   Units
PathUtilsBenchmark.globMatching_UNIX              thrpt    2  1.406          ops/us
PathUtilsBenchmark.globMatching_UNIX_INVERTED     thrpt    2  1.405          ops/us
PathUtilsBenchmark.globMatching_WINDOWS           thrpt    2  1.410          ops/us
PathUtilsBenchmark.globMatching_WINDOWS_INVERTED  thrpt    2  1.383          ops/us


AFTER: (StringTokenizer)
Benchmark                                          Mode  Cnt  Score   Error   Units
PathUtilsBenchmark.globMatching_UNIX              thrpt    2  1.212          ops/us
PathUtilsBenchmark.globMatching_UNIX_INVERTED     thrpt    2  1.220          ops/us
PathUtilsBenchmark.globMatching_WINDOWS           thrpt    2  1.030          ops/us
PathUtilsBenchmark.globMatching_WINDOWS_INVERTED  thrpt    2  1.090          ops/us
```

This leads me to believe that the current `charAt` implementation is the most consistently performant solution.

Reply:
@knutwannheden, is there anything else that you'd like to see here before this is merged?

Reply:
No, LGTM. Thanks!